### PR TITLE
Add jwks endpoint for return realm/key JSON encoded keys

### LINKED ISSUE #93: Implement JWKS discovery for verification certificate keys
### TL;DR

In order to facilitate key rotation and automatic discovery by an exposure notifications server, implement a JWKS document endpoint that provides the current public key material.

### Design

**Proposal**

Needs to be a well known - unauthenticated endpoint for retrieval.

the KIDs need to be encoded in such a way that they can convey multi-tenancy of the verification server.

For example, a kid might look like "us-wa-00001" 

See example JWK here -> https://tools.ietf.org/html/rfc7517#section-3

/kind enhancement
/assign

### PR DESCRIPTION
**Pull request by jeremyfaller**
Fixes #93

## Proposed Changes

* Add /jwks/XX/XX endpoint for key collection.
* Fix typo
*

**Release Note**

```release-note
NONE
```

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -304,6 +305,18 @@ func realMain(ctx context.Context) error {
...
+	// jwks
+	{
+		jwksSub := r.PathPrefix("/jwks").Subrouter()
+		jwksSub.Use(rateLimit)
+
+		jwksController, err := jwks.New(ctx, db, cacher, h)
+		if err != nil {
+			return fmt.Errorf("failed to create jwks controller: %w", err)
+		}
+		jwksSub.Handle("/{realm}/{keyid}", jwksController.HandleIndex()).Methods("GET")
```
Comment (mikehelmick) [inline cmd/server/main.go:317]: The key id doesn't need to be in the URL - a single endpoint should list all non-destroyed keys for a realm.
That way the exposure-notificiations-server has a single document to point to.

Sorry if our chat conversation steered you wrong here.

Comment (jeremyfaller) [reply cmd/server/main.go:317]: It did. I'll fix and reupload.

Comment (jeremyfaller) [reply cmd/server/main.go]: Ok, this might be closer to what you had in mind. LMK.

**Comment by jeremyfaller** [comment]
The conflict is easy enough to edit up here, so I'm not going to bother syncing client, etc.... I'm happy to do it if this LGTY.

File: pkg/controller/jwks/jwks.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+				return nil, err
+			}
+			return keys, nil
+		}); err != nil {
+			c.renderError(w, err)
+			return
+		}
+
+		// Get the keys.
+		encoded := make([]*jwk.JWK, len(keys))
```
Comment (mikehelmick) [inline pkg/controller/jwks/jwks.go:77]: does it make sense for this to be the cached entity?

File: pkg/controller/jwks/jwks.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	var realm *database.Realm
+	realm, err = c.db.FindRealm(uint(realmID))
+	if err != nil {
+		return nil, errBadRealm
+	}
+	return realm, nil
+}
+
+// reportError renders an error.
+func (c *Controller) renderError(w http.ResponseWriter, err error) {
```
Comment (mikehelmick) [inline pkg/controller/jwks/jwks.go:113]: there are existing helpers for this already

File: pkg/database/realm.go [hunk 4]
```diff
@@ -249,7 +249,7 @@ func (r *Realm) SMSProvider(db *Database) (sms.Provider, error) {
 
 // GetCurrentSigningKey returns the currently active signing key, the one marked
 // active in the database. If there is more than one active, the most recently
-// crated on wins. Should not occur due to transactional update.
+// created on wins. Should not occur due to transactional update.
```
Comment (mikehelmick) [inline pkg/database/realm.go:252]: while you're here, s/on/one/

File: cmd/server/assets/realmkeys.html [hunk 5]
```diff
@@ -93,7 +93,7 @@
                 {{range $rk := .realmKeys}}
                 <tr>
                   <td>
-                    <span class="text-monospace">{{$rk.GetKID}}</span>
+                    <a href="/jwks/{{$rk.RealmID}}/{{$rk.ID}}" class="text-monospace">{{$rk.GetKID}}</a>
```
[the remark this answers is not in the document]
Comment (jeremyfaller) [reply cmd/server/assets/realmkeys.html:96]: Yup. my mistake after previous convo on this PR.

File: pkg/controller/jwks/jwks.go (hunk 3 above)
Comment (jeremyfaller) [reply pkg/controller/jwks/jwks.go:113]: Done.

File: pkg/controller/jwks/jwks.go (hunk 2 above)
Comment (jeremyfaller) [reply pkg/controller/jwks/jwks.go:77]: Arbitrary to me. I felt the rest of the code was so minimal, and used its own caches that it wasn't a burden. Updated.

File: pkg/database/realm.go (hunk 4 above)
Comment (jeremyfaller) [reply pkg/database/realm.go:252]: Ha. Didn't even finish the sentence. Saw the first typo, and stopped.

**Pull request merged**